scrapy/core/downloader/__init__.py

```diff
@@ -150,13 +150,30 @@ def _process_queue(self, spider, slot):
         while slot.queue and slot.free_transfer_slots() > 0:
             slot.lastseen = now
             request, deferred = slot.queue.popleft()
+
+            # Handling request delay set to individual requests
+            if request.meta.get('request_delay'):
+                request_delay = request.meta.get('request_delay')
+                reactor.callLater(
+                    request_delay,
+                    self._handle_request_delay,
+                    slot,
+                    request,
+                    spider,
+                    deferred)
+                continue
```

Comment:
1. I don't think that it is a good idea to add additional layer of deferreds to implement this.
I think that relatively safe way to implement this is to somehow affect to variables that calculate delay:
https://github.com/scrapy/scrapy/blob/26836c4e1ae9588ee173c5977fc6611364ca7cc7/scrapy/core/downloader/__init__.py#L142-L147
Results of my previous observartions of this part of code I posted in this https://github.com/scrapy/scrapy/issues/802#issuecomment-500245345

In my https://github.com/scrapy/scrapy/pull/3835 pull request I changed downloader slot `delay` method ( it return delay from `meta` parameter instead of regular delay from `DOWNLOAD_DELAY` setting)

2. You didn't make changes to original delay logic .It means that real delay will be:
regular delay from `DOWNLOAD_DELAY` + delay from 'request_delay' meta.
It is also good but I am not sure that it is expected (especially for non zero `DOWNLOAD_DELAY`)

Reply:
@GeorgeA92
1. The slot mechanism doesn't handle individual request delay and hence I thought this is needed 
2. The solution in your PR doesn't give the intended result, I've added an explanation here https://github.com/scrapy/scrapy/pull/3835#discussion_r571984220
Change in this PR is to delay the individual request and your PR has another solution which delays the entire slot depending upon delay settings on individual request.

Reply:
Hey @realslimshanky! Nice to see some progress on #802 :)

I think the implementation may need to be a a it more convoluted. 
In addition to what @GeorgeA92 said, consider the following. 

For example, imagine you're sending 16 requests with request_delay = 1 day, and many other requests, without request_delay. Eventually all these 16 requests will go from scheduler queue to the downloader queue. With default Scrapy's global concurrency (16), it'd mean that the whole spider is blockeed for 1 day, even if there are other requests which can be processed in the meantime.

So it seems to implement the delay in an efficient way, scheduler needs to be involved in some way.
I think request_delay semantics is really important here. With the current implementation it can be a bit hard to explain: you're making a deferred call after request is fetched from the downloader

Reply:
@kmike I tested it out, the requests are not blocked since the implementation is done to delay the `_download` which is independent of all other requests. So, let's say you schedule these URLs with concurrency set to 8
```python
CONCURRENT_REQUESTS = 8
CONCURRENT_REQUESTS_PER_DOMAIN = 8

def start_requests(self):
        urls = [
            'http://quotes.toscrape.com/page/1/',
            'http://quotes.toscrape.com/page/2/',
            'http://quotes.toscrape.com/page/3/',
            'http://quotes.toscrape.com/page/4/',
            'http://quotes.toscrape.com/page/5/',
            'http://quotes.toscrape.com/page/6/',
            'http://quotes.toscrape.com/page/7/',
            'http://quotes.toscrape.com/page/8/',
            'http://quotes.toscrape.com/page/9/',
            'http://quotes.toscrape.com/page/10/',
            'http://quotes.toscrape.com/page/11/',
            'http://quotes.toscrape.com/page/12/',
            'http://quotes.toscrape.com/page/13/',
            'http://quotes.toscrape.com/page/14/',
            'http://quotes.toscrape.com/page/15/',
            'http://quotes.toscrape.com/page/16/',
            'http://quotes.toscrape.com/page/17/',
            'http://quotes.toscrape.com/page/18/',
            'http://quotes.toscrape.com/page/19/',
        ]
        for index, url in enumerate(urls, 1):
            if index == 1 :
                yield scrapy.Request(url=url, callback=self.parse, meta={'request_delay': 10})
            if index == 2 :
                yield scrapy.Request(url=url, callback=self.parse, meta={'request_delay': 10})
            else:
                yield scrapy.Request(url=url, callback=self.parse)

def parse(self, response):
        print(f'Parsed {response.url}')
```

The order of execution would be
```
2021-02-08 18:35:56 [scrapy.core.engine] DEBUG: Crawled (404) <GET http://quotes.toscrape.com/robots.txt> (referer: None)
2021-02-08 18:35:57 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/3/> (referer: None)
Parsed http://quotes.toscrape.com/page/3/
2021-02-08 18:35:57 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/9/> (referer: None)
2021-02-08 18:35:57 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/8/> (referer: None)
2021-02-08 18:35:57 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/4/> (referer: None)
2021-02-08 18:35:57 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/7/> (referer: None)
2021-02-08 18:35:57 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/5/> (referer: None)
2021-02-08 18:35:57 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/6/> (referer: None)
Parsed http://quotes.toscrape.com/page/9/
Parsed http://quotes.toscrape.com/page/8/
Parsed http://quotes.toscrape.com/page/4/
Parsed http://quotes.toscrape.com/page/7/
Parsed http://quotes.toscrape.com/page/5/
Parsed http://quotes.toscrape.com/page/6/
2021-02-08 18:35:57 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/12/> (referer: None)
2021-02-08 18:35:57 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/14/> (referer: None)
2021-02-08 18:35:57 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/11/> (referer: None)
2021-02-08 18:35:57 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/13/> (referer: None)
2021-02-08 18:35:57 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/15/> (referer: None)
2021-02-08 18:35:57 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/10/> (referer: None)
Parsed http://quotes.toscrape.com/page/12/
Parsed http://quotes.toscrape.com/page/14/
Parsed http://quotes.toscrape.com/page/11/
Parsed http://quotes.toscrape.com/page/13/
Parsed http://quotes.toscrape.com/page/15/
Parsed http://quotes.toscrape.com/page/10/
2021-02-08 18:35:58 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/16/> (referer: None)
2021-02-08 18:35:58 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/17/> (referer: None)
2021-02-08 18:35:58 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/18/> (referer: None)
2021-02-08 18:35:58 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/19/> (referer: None)
Parsed http://quotes.toscrape.com/page/16/
Parsed http://quotes.toscrape.com/page/17/
Parsed http://quotes.toscrape.com/page/18/
Parsed http://quotes.toscrape.com/page/19/
2021-02-08 18:36:07 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/2/> (referer: None)
2021-02-08 18:36:07 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/page/1/> (referer: None)
Parsed http://quotes.toscrape.com/page/2/
Parsed http://quotes.toscrape.com/page/1/
```

So, the normal operation of the spider will continue, only the download of the 2 requests will be delayed and the response is sent to callback once it's done.

Reply:
@realslimshanky but imagine that you have `{'request_delay': 10000}` for 16 requests, and you have 16K requests in start_urls - it would block the downloader, meaning that at some point all slots would be waiting for 10000s, and no other requests will be processed during this time. This is not about event loop blocking, but more about scrapy overall behavior.

Reply:
@kmike I get that if we increase request_delay too much it'll wait for said amount of time before finishing execution. But I don't understand how it'll block other requests from processing?

Reply:
Downloader has a limit on a number of requests are active at the same time (CONCURRENT_REQUESTS option). In the way delay is implemented, all requests which are being delayed are considered active, i.e. they're taking slots in the downloader. 

It means that if you have CONCURRENT_REQUESTS requests with a large request_delay, all slots in downloader get busy, and so Scrapy stops doing anything else. 

I think we can do better than this: if there are requests waiting for request_delay, why not do other requests during this time? 

To make it easier to think about, consider a case with CONCURRENT_REQUESTS=1 and a big request_delay values (in hours, not seconds). How can we make it work? 

A related thing is that request_delay happens only after request gets to the downloader. So e.g. a request with request_delay=5m was put to the Scheduler 1 hour ago. It gets to the Downloader after this 1hr, but it is still delayed for 5m more. 

This may not match user experience; we would need to explain very clearly what request_delay means.

Options:

1. Min time between `yield scrapy.Request` and a time a download starts
2. Min time between

Reply:
@kmike I understand it now, specially with the `CONCURRENT_REQUESTS` setting set to 1, that's a great example for this scenario.
Okay, so instead of intercepting requests on the downloader, scheduler would be a good place. And also, we need a time aware scheduler to handle request_delay and remove redundancies.
I'll think over it and see what can be done to resolve it.

Reply:
@kmike 

> For example, imagine you're sending 16 requests with request_delay = 1 day, and many other requests, without request_delay. Eventually all these 16 requests will go from scheduler queue to the downloader queue. With default Scrapy's global concurrency (16), it'd mean that the whole spider is blocked for 1 day, even if there are other requests which can be processed in the meantime.

If it is expected that downloader slot should be free to process zero delay requests (process them before 1day delay requests)
then that 1day delay requests definitely should be delayed from scheduler side.